### PR TITLE
Introduce Runtime Events

### PR DESCRIPTION
**Pull request by gavofyork**
Based on #570 

Runtime events are roughly equivalent to "Events" in Ethereum ABI model. They're high-level notifications for particular conditions hit by the runtime. Examples of such conditions might be a session rotation or a slash. When emitted, they can take arguments to further describe the event; an example for a `Slash` event would be the validator ID (`AccountId`) and the amount slashed (`Balance`).

Substrate has no specific means of encoding events into the block header like Ethereum's `TransactionReceipt`. Rather records of events are collated by the `System` and `Executive` modules so they form a vector of `EventRecord`s (which includes metadata of the event; right now that's just which extrinsic that caused it to be emitted) and left in storage. An API is exposed from the runtime to query this storage item and return the events during the block. The storage item is cleared during the initialisation of the following block.

It is assumed that middleware or a client module on the node side will query new blocks and populate an indexed database with each block's events. Since the event records are stored as a simple storage item, light-clients can track and monitor events in a brute-force fashion. However in this PR, no effort is made to index events into storage to make the monitoring of specific parameters efficient for light-clients. Further consideration is required and can be introduced at a later stage.

Events are introduced into the runtime in much the same way as dispatching. Each module that wants to be able to deposit events exposes an `Event` type, usually an `enum`, which must implement `Encode`, `Decode`, `Clone`, `Eq` and `Debug`. The module's configuration trait is required to be given an "outer" `Event` type which allows the module's type to be converted `Into` the event type expected by the system module's `deposit_event` function (which is the same as the "outer" event type, but we assume a `From` relationship to avoid a cyclic type dependency). This was the most ergonomic, least round-about way I could find of getting the job done.

In the runtime, a new macro is used `impl_outer_event!` which is passed each of the modules that emit events and creates a "super-enum" type that each of the types can be converted `Into`. This type is then passed into each module's configuration trait as the outer event type.

Yet to do:
- [x] Unit tests.
- [x] Integration style tests.
- Light-client friendly event indexing (i.e. index events into storage according to their parameters). Another PR.

### CODE REVIEW

File: substrate/runtime/staking/src/lib.rs [hunk 1]
```diff
@@ -942,27 +989,38 @@ impl<T: Trait> Module<T> {
...
+
 	/// Kill an account's free portion.
 	fn on_free_too_low(who: &T::AccountId) {
 		Self::decrease_total_stake_by(Self::free_balance(who));
 		<FreeBalance<T>>::remove(who);
 		<Bondage<T>>::remove(who);
-		T::OnAccountKill::on_account_kill(who);
+
+		// TODO: rename this to on_free_account_kill or some such, since it only pertains to removing the 
+		// free-balance part of the account, not the whole thing.
```
Comment (pepyakin) [inline substrate/runtime/staking/src/lib.rs]: I guess this is sorted now?

File: substrate/runtime/primitives/src/traits.rs [hunk 2]
```diff
@@ -104,6 +104,14 @@ pub struct Identity;
 impl<T> Convert<T, T> for Identity {
 	fn convert(a: T) -> T { a }
 }
+pub struct Empty;
+impl<T> Convert<T, ()> for Empty {
+	fn convert(_: T) -> () { () }
+}
+pub struct UseInto;
```
Comment (pepyakin) [inline substrate/runtime/primitives/src/traits.rs]: How these are expected to be used? I can't see these used anywhere

File: README.adoc [hunk 3]
```diff
@@ -4,9 +4,87 @@ Next-generation framework for blockchain innovation.
...
+Substrate chains have three distinct features that make them "next-generation": a dynamic, self-defining state-transition function, light-client functionality from day one and a progressive consensus algorithm with fast block production and adaptive, definite finality. The STF, encoded in WebAssembly, is known as the "runtime". This defines the `execute_block` function, and can specify everything from the staking algorithm, transaction semantics, logging mechanisms and procedures for replacing any aspect of itself or of the blockchain's state ("governance"). Because the runtime is entirely dynamic all of these can be switched out or upgraded at any time. A Substrate chain is very much a "living organism".
```
Comment (pepyakin) [inline README.adoc]: Can we expand what is STF little earlier, like `a dynamic, self-defining state-transition function (STF)`?

**Comment by gavofyork** [comment]
In reality all runtime code is trusted. There's no need to call into system module to clear or change events - you could just alter storage directly with the `runtime_io` module. The smart-contract sandbox is the only way of ensuring that (untrusted) code can't do stuff like clear bits of storage.

File: README.adoc (hunk 3 above)
Comment (gavofyork) [reply README.adoc]: Yeah the docs is a WiP; I'll continue working on it, but I figured it's better to have something there sooner than later.

**Pull request merged**